File: ios/Classes/TiGooglemapsIndoorDisplayProxy.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2017 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsIndoorDisplayProxy.h"
#import "TiGooglemapsIndoorLevelProxy.h"

@implementation TiGooglemapsIndoorDisplayProxy

- (id)_initWithPageContext:(id<TiEvaluator>)context andIndoorDisplay:(GMSIndoorDisplay *)indoorDisplay
{
    if (self = [super _initWithPageContext:pageContext]) {
        _indoorDisplay = indoorDisplay;
        [_indoorDisplay setDelegate:self];
    }
    
    return self;
}

#pragma mark Public API

- (id)activeBuilding
{
    return [self dictionaryFromIndoorBuilding:[_indoorDisplay activeBuilding]];
}

- (id)activeLevel
{
    return [[TiGooglemapsIndoorLevelProxy alloc] _initWithPageContext:[self pageContext] andIndoorDisplay:[_indoorDisplay activeLevel]];
}

- (void)setActiveLevel:(id)activeLevel
{
    ENSURE_UI_THREAD(setActiveLevel, activeLevel);
    ENSURE_TYPE(activeLevel, TiGooglemapsIndoorLevelProxy);
    
    [_indoorDisplay setActiveLevel:[(TiGooglemapsIndoorLevelProxy *)activeLevel indoorLevel]];
}

#pragma mark GMSIndoorDisplayDelegate

- (void)didChangeActiveLevel:(GMSIndoorLevel *)level
{
    if ([self _hasListeners:@"didChangeActiveLevel"]) {
        [self fireEvent:@"didChangeActiveLevel" withObject:[[TiGooglemapsIndoorLevelProxy alloc] _initWithPageContext:[self pageContext]
                                                                                                     andIndoorDisplay:level]];
    }
}

- (void)didChangeActiveBuilding:(GMSIndoorBuilding *)building
{
    if ([self _hasListeners:@"didChangeActiveBuilding"]) {
        [self fireEvent:@"didChangeActiveBuilding" withObject:[self dictionaryFromIndoorBuilding:building]];
    }
}

#pragma mark Utilities

- (NSDictionary<NSString *, id> *)dictionaryFromIndoorBuilding:(GMSIndoorBuilding *)indoorBuilding
{
    NSMutableArray *levels = [NSMutableArray arrayWithCapacity:[[indoorBuilding levels] count]];
    
    for (GMSIndoorLevel *indoorLevel in indoorBuilding.levels) {
        [levels addObject:[[TiGooglemapsIndoorLevelProxy alloc] _initWithPageContext:[self pageContext]
                                                                    andIndoorDisplay:indoorLevel]];
    }
    
    return @{
        @"defaultLevelIndex": NUMUINTEGER([indoorBuilding defaultLevelIndex]),
        @"isUnderground": NUMBOOL([indoorBuilding isUnderground]),
        @"levels": levels
    };
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsIndoorDisplayProxy.swift
/**
 * Axway Titanium
 */

import Foundation
import TitaniumKit
import GoogleMaps

@objc(TiGooglemapsIndoorDisplayProxy)
public class TiGooglemapsIndoorDisplayProxy: TiProxy, GMSIndoorDisplayDelegate {
  private var indoorDisplay: GMSIndoorDisplay?
  
  @objc(_initWithPageContext:andIndoorDisplay:)
  public func _init(withPageContext context: TiEvaluator?, andIndoorDisplay display: GMSIndoorDisplay) -> Self? {
    guard let proxy = super._init(withPageContext: context) else { return nil }
    indoorDisplay = display
    indoorDisplay?.delegate = self
    return proxy
  }
  
  @objc public func activeBuilding() -> [String: Any]? {
    guard let building = indoorDisplay?.activeBuilding else { return nil }
    return dictionary(from: building)
  }
  
  @objc public func activeLevel() -> TiGooglemapsIndoorLevelProxy? {
    guard let level = indoorDisplay?.activeLevel else { return nil }

    return TiGooglemapsIndoorLevelProxy()._init(withPageContext: pageContext, andIndoorLevel: level)
  }
  
  @objc public func setActiveLevel(_ value: TiGooglemapsIndoorLevelProxy?) {
    guard let level = value?.indoorLevel() else { return }
    indoorDisplay?.activeLevel = level
  }
  
  // MARK: - GMSIndoorDisplayDelegate
  
  public func didChangeActiveLevel(_ indoorLevel: GMSIndoorLevel?)
  {
    guard _hasListeners("didChangeActiveLevel"), let indoorLevel else { return }
    
    let levelProxy = TiGooglemapsIndoorLevelProxy()._init(withPageContext: pageContext, andIndoorLevel: indoorLevel)
    fireEvent("didChangeActiveLevel", with: ["level": levelProxy as Any])
  }
  
  public func didChangeActiveBuilding(_ building: GMSIndoorBuilding?) {
    guard _hasListeners("didChangeActiveBuilding"), let building else { return }
    fireEvent("didChangeActiveBuilding", with: dictionary(from: building))
  }
  
  // MARK: - Utilities
  
  private func dictionary(from building: GMSIndoorBuilding) -> [String: Any] {
    let levels: [Any] = building.levels.compactMap { level in
      return TiGooglemapsIndoorLevelProxy()._init(withPageContext: pageContext, andIndoorLevel: level)
    }
    
    return [
      "defaultLevelIndex": building.defaultLevelIndex,
      "isUnderground": building.isUnderground,
      "levels": levels
    ]
  }
}
